File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/content/PseudoTextFold.swift
//
//  PseudoTextFold.swift
//  StyleEngine/content — wave 42, lane W2 (the iOS pseudos seam);
//  wave 43, lane V7 (the styled-text extension).
//
//  The COMPONENT half of the ordinary-element generated-content path:
//  fold the baked ::before text in as a LEADING inline run and the baked
//  ::after text as a TRAILING one, by rewriting the component's `text`
//  before the renderer reads it. CSS 2.1 §12.1 places ::before content
//  immediately before, and ::after immediately after, the element's own
//  document-tree content; web's NodeRenderer spells the same order as
//  positional children (marker, before, text, children, after —
//  runtimes/web/src/renderer/NodeRenderer.ts). Concatenating
//  before + own text + after reproduces that glyph order exactly for the
//  foldable population, and because `component.text` is the ONE channel
//  every downstream consumer reads (leaf PlaceholderLabel, the
//  mixed-content leading label, wrap-width measurement), paint and
//  measurement agree by construction — no second text channel to drift.
//
//  A pseudo inherits its typography from the originating element
//  (css-pseudo-4 §2; web's <span> inherits the same way), so an UNSTYLED
//  fold renders with precisely the style the host's text run already has.
//  Wave 43 adds the bucket's OWN styling (PseudoTextBridge → typed
//  properties): `component.text` is ONE uniform run, so a styled fold is
//  only honest when the pseudo is the component's SOLE ink — then the
//  typed entries append to the copy's property list (LAST, so they win
//  the extractors' last-wins cascade like the pseudo's own declaration
//  beating the host's inherited value) and the ordinary text pipeline
//  paints them. Any other styled shape refuses, named.
//

import Foundation

/// The pre-render fold: `pseudos.{before,after}._text` → `component.text`.
// enum: pure namespace — never instantiated (mirrors PseudoTextBridge).
enum PseudoTextFold {

    /// Rewrite `component` with its foldable pseudo text folded into
    /// `text`; identity (the SAME value back, field-for-field) whenever
    /// there is nothing to fold — which is every component of the
    /// committed 327-fixture baseline corpus (none carries `pseudos`),
    /// keeping all committed captures byte-stable.
    static func resolve(_ component: IRComponent) -> IRComponent {
        // No bucket at all → the overwhelmingly common identity path.
        guard let pseudos = component.pseudos else { return component }
        // The body-root's bucket belongs to the RootPseudoBox BOX path
        // (wave-28 lane PG): that family is `content:"" + width/height/
        // background`, a box not a text run, and folding here would
        // double-render anything RootPseudoBoxView paints. Same guard
        // direction as RootPseudoBox's banner ("this file must never
        // double-render either"), from the opposite shore.
        guard component.meta?.role != "body-root" else { return component }
        // `pseudos.marker` belongs to the list machinery (meta.markerText
        // and the markerPlacement loop in ComponentRenderer) — folding it
        // into `text` would double-paint against that path. Named once,
        // per the no-silent-fallthrough rule, not per component.
        if pseudos["marker"] != nil {
            PropertyTracker.logOnce(
                key: "pseudotext-marker",
                message: "[PseudoText] pseudos.marker is delegated to the list marker path — not folded")
        }
        // `meta.runs` is AUTHORITATIVE over `text` (spec 03 §4.1): when a
        // run plan resolves, the renderer drops the text slot entirely, so
        // a fold into `text` would silently vanish. Refuse and name it.
        if component.meta?.runs != nil {
            // Only worth a log line when there was something to lose.
            if pseudos["before"] != nil || pseudos["after"] != nil {
                PropertyTracker.logOnce(
                    key: "pseudotext-runs",
                    message: "[PseudoText] component carries meta.runs — pseudo text not folded (runs own the content slot)")
            }
            return component
        }
        // The em/% base for the pseudo's own `font-size`: css-values-4
        // §5.1.1 resolves it against the INHERITED size, and the pseudo's
        // parent IS the originating element — so the host's own declared
        // size (last FontSize wins, the extractors' cascade) or the 16px
        // document default when the host declares none (the only base the
        // wave-43 corpus produces: contain-content-011's host is bare).
        let emBasePx = component.properties.last { $0.type == "FontSize" }
            .flatMap { ValueExtractors.extractPx($0.data) }.map(Double.init) ?? 16.0
        // ── Wave-49 lane A2: the generated-BOX path claims FIRST ─────────
        // A bucket declaring a block-level box is not an inline run at all
        // (CSS 2.1 §12.1 generates it INSIDE the element as its first/last
        // child), so PseudoBoxBridge takes it and the text bridge below must
        // not also see it — one claim function, two consumers, so the same
        // content can never render twice. nil for every bucket the box
        // bridge does not own, which is every wave-48 payload except
        // css-pseudo/before-as-flex-container's `display: flex` ::before.
        let hostRole = component.meta?.role
        let beforeBox = PseudoBoxBridge.claim(
            bucket: pseudos["before"], role: "before", hostRole: hostRole)
        let afterBox = PseudoBoxBridge.claim(
            bucket: pseudos["after"], role: "after", hostRole: hostRole)
        // ::before — always foldable as the LEADING run: web renders its
        // span first, then the element's own text, whether or not real
        // children follow (NodeRenderer's positional order).
        var before = beforeBox != nil ? nil : PseudoTextBridge.inlineRun(
            bucket: pseudos["before"], role: "before", emBasePx: emBasePx)
        // ::after — foldable only on a CHILDLESS component: with composed
        // children, web paints the after-span BEHIND them (text, children,
        // after), an order a leading text fold cannot express.
        var after: PseudoTextBridge.PseudoInlineRun? = nil
        // Bucket presence checked first so the refusal log fires only for
        // components that actually carry an ::after payload.
        if let afterBucket = pseudos["after"], afterBox == nil {
            // nil/empty children = the leaf shape (never [], per the
            // IRComponent decode contract) — the safe append position.
            if component.children?.isEmpty ?? true {
                after = PseudoTextBridge.inlineRun(
                    bucket: afterBucket, role: "after", emBasePx: emBasePx)
            } else {
                // Named refusal: the trailing seam behind a child stack is
                // a renderer change this fold deliberately does not make.
                PropertyTracker.logOnce(
                    key: "pseudotext-after-children",
                    message: "[PseudoText] ::after on a component with children is not folded (it must trail the children)")
            }
        }
        // ── the wave-43 uniformity gate ─────────────────────────────────
        // `component.text` is ONE uniform run, so typed styling may only
        // apply when the styled pseudo is the SOLE ink: host text empty
        // AND the opposite side folding nothing. Judged against the
        // PRE-GATE pair so two styled sides refuse each other
        // symmetrically instead of the later one stealing the slot.
        let hostText = component.text ?? ""
        // Snapshot of the pre-gate state both checks read.
        let beforeFolds = before != nil, afterFolds = after != nil
        // A styled ::before sharing the run with host text or an ::after.
        if let b = before, !b.styling.isEmpty, !(hostText.isEmpty && !afterFolds) {
            // Named refusal — the declared style cannot be honoured
            // uniformly, and painting it in the host's style would lie.
            PropertyTracker.logOnce(
                key: "pseudotext-styled-nonuniform-before",
                message: "[PseudoText] ::before styling cannot apply — the fold is not the component's sole text run")
            before = nil
        }
        // Mirror check for a styled ::after (census: not present today).
        if let a = after, !a.styling.isEmpty, !(hostText.isEmpty && !beforeFolds) {
            PropertyTracker.logOnce(
                key: "pseudotext-styled-nonuniform-after",
                message: "[PseudoText] ::after styling cannot apply — the fold is not the component's sole text run")
            after = nil
        }
        // Nothing survived the gates → identity, so the view tree of every
        // non-foldable component is untouched. Wave-49 lane A2: a claimed
        // BOX counts as something to fold even when no text run does.
        guard before != nil || after != nil || beforeBox != nil || afterBox != nil else {
            return component
        }
        // The fold itself: before + own text + after, in CSS 2.1 §12.1
        // order. `_text` bakes its own separators ("1 " / " 1"), so plain
        // concatenation is the whole job — non-empty by construction
        // because the bridge never returns an empty string.
        let folded = (before?.text ?? "") + hostText + (after?.text ?? "")
        // The surviving styling (at most one side by the gate above) —
        // appended AFTER the host's properties so the pseudo's own
        // declarations win the extractors' last-wins cascade.
        let styling = (before?.styling ?? []) + (after?.styling ?? [])
        // Wave-49 lane A2: when ONLY a box was claimed no text run folded,
        // so the host's own `text` must ride through VERBATIM — nil and ""
        // are different values in the decode contract ("absent" vs
        // "extracted, was empty"), and `folded` would flatten nil to "".
        let resolvedText = (before == nil && after == nil) ? component.text : folded
        // Field-for-field copy with `text` rewritten and the typed styling
        // appended. `pseudos` is KEPT on the copy: ContentsUnboxing's
        // eligibility gate reads it (a pseudos-carrying component never
        // unboxes) and spec 05 rule 4 wants the wire shape re-derivable —
        // consuming is not erasing.
        return IRComponent(id: component.id, name: component.name,
                           properties: component.properties + styling,
                           selectors: component.selectors,
                           media: component.media,
                           children: splicedChildren(component, beforeBox, afterBox),
                           slot: component.slot,
                           text: resolvedText,
                           pseudos: component.pseudos,
                           meta: component.meta,
                           variables: component.variables)
    }

    /// The child list with the claimed generated boxes spliced in — CSS 2.1
    /// §12.1 order: ::before, the element's own children, ::after. Returns
    /// the ORIGINAL array (nil included) when no box was claimed, so the
    /// text-only fold stays byte-identical to wave 48.
    private static func splicedChildren(
        _ host: IRComponent,
        _ beforeBox: PseudoBoxBridge.BoxClaim?,
        _ afterBox: PseudoBoxBridge.BoxClaim?
    ) -> [IRComponent]? {
        guard beforeBox != nil || afterBox != nil else { return host.children }
        var kids: [IRComponent] = []
        if let b = beforeBox { kids.append(boxComponent(host, "before", b)) }
        kids.append(contentsOf: host.children ?? [])
        if let a = afterBox { kids.append(boxComponent(host, "after", a)) }
        return kids
    }

    /// One generated box as a synthetic child component.
    ///
    /// Making it a REAL child is the whole fix: the host's own layout
    /// branch (block flow here, flex/grid elsewhere) then places it as the
    /// first/last in-flow child exactly like an authored child, and
    /// StyleBuilder paints its typed declarations through the one existing
    /// pipeline — no view-tree plumbing anywhere else.
    ///
    /// `slot` is nil on purpose: it is COMPOSER input (spec 03 — only a
    /// composer reads it, and composition has already happened by the time
    /// this fold runs). `pseudos` is nil too — a generated box has no
    /// generated content of its own, which also makes the fold structurally
    /// non-recursive.
    private static func boxComponent(
        _ host: IRComponent, _ role: String, _ claim: PseudoBoxBridge.BoxClaim
    ) -> IRComponent {
        IRComponent(
            // Stable, collision-free identity: the CSS selector spelling of
            // the box. SwiftUI keys the ForEach by it, so it must not move
            // between renders of the same host.
            id: "\(host.id)::\(role)", name: "\(host.name)::\(role)",
            properties: claim.properties,
            selectors: nil, media: nil, children: nil, slot: nil,
            // The literal/baked content string; nil rather than "" so the
            // renderer's has-text tests read the same as an authored empty
            // leaf (`content: ""` boxes paint their background only).
            text: claim.text.isEmpty ? nil : claim.text,
            pseudos: nil, meta: nil)
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/content/PseudoTextFold.kt
package com.styleconverter.runtime.content

// PseudoTextFold — wave-50 lane B3: the Compose twin of iOS
// StyleEngine/content/PseudoTextFold.swift, for the ::after half only.
//
// THE MEASURED DEFECT (wave49-final captures, Android column). Compose
// composes an ordinary element's generated content as a WRAPPER AROUND the
// host — ContentApplier.ContentWithPseudoElements builds `Row { before,
// content, after }` — while CSS 2.1 §12.1 / css-pseudo-4 §4.1 put the
// ::after box INSIDE the originating element, as its LAST child (web spells
// exactly that: NodeRenderer emits the pseudo span as a positional child;
// iOS folds it into the host's text run). Outside the host box the ::after
// has to share the line with the host's own block box, and in composed-WPT
// capture that box takes `Modifier.fillMaxWidth()` (ComponentRenderer's
// blockFlowWidth — CSS 2.1 §10.3.3 stretch fit / css-sizing-3 §5 used size).
// A Compose Row measures unweighted children in order against the space its
// predecessors left, so the host consumes the whole line and the ::after Text
// is measured at maxWidth = 0: it still occupies the row (its zero-width
// layout wraps one grapheme per line) but draws NOTHING.
//
// Three capture signatures of that one cause, all in
// tools/titan/runs/wave49-final/sections/:
//   * css-lists/counter-reset-reversed-pseudo-001 — Android paints "B7"
//     and "B4" (the ::before runs) and loses "A5"/"A2"; the ink bands sit
//     at y20-31 and y60-71, a 40px row pitch, where iOS has them 20px apart
//     at y20-31 / y40-51. The doubled pitch IS the zero-width wrap of the
//     two-glyph ::after — proof the run was composed and simply drew
//     nothing (android f 0.9884 vs iOS P 0.9999).
//   * css-counter-styles/cjk-decimal/counter-cjk-decimal — every component
//     is `::after`-ONLY, so the Android capture is BLANK in the literal
//     sense: the whole 390x696 PNG decodes to ONE colour,
//     RGBA(255,255,255,255), with no ink anywhere
//     (tools/titan/results/wave50-S6/blank-capture-census.txt). That is the
//     strongest evidence for the defect. The coverage number beside it was
//     misquoted until wave-50 lane F4 (skeptic S6-18): the android-ref row's
//     semantic presence is aCoveragePct 0.000 — exact, the capture has no
//     ink — against bCoveragePct 0.816, the REF side as measured in that
//     row's own 696-tall frame; 0.946 is the web-ref / ios-ref rows' ref
//     figure and belongs to them, not here. The canvas grows 600 → 696 px on
//     the wrapped zero-width rows (android f 0.9407 · iOS P 0.9934).
//   * css-contain/contain-content-011 — the host declares `width: 100px`,
//     so there is no squeeze and the defect shows as PLACEMENT instead: the
//     "25" is painted at x≈118, to the RIGHT of the host's border box,
//     where web and iOS put it at x≈16 inside it.
//
// SCOPE — ::after ONLY, deliberately. The ::before run is composed FIRST in
// the same Row, so it already lands adjacent to the host's content edge and
// renders correctly today; folding it too would move 55 corpus tests that
// carry a before-only bucket, which is a device-A/B change, not a lane
// change. The ::after side has no such luck: it is the one the stretch-fit
// host squeezes. Blast radius of this fold, replayed over all 1435
// wave49-final per-test-ir documents with the gates in this file and in
// PseudoTextBridge: 13 tests, 78 components — of which 9 Android cells fail
// today and 4 pass (tools/titan/results/wave50-B3/_note.md has the table).
//
// ONE CLAIM, TWO CONSUMERS: a bucket this fold takes must never ALSO render
// through ContentApplier's wrapper, so [resolve] reports its decision in
// [Folded.afterFolded] and the renderer passes it to
// PseudoBucketExtractor.extractBeforeAfterConfig — the same discipline
// PseudoBoxFold and PseudoBucketExtractor share through
// PseudoGeneratedBox.claim. The decision is computed ONCE, beside the fold,
// because the fold rewrites `_text` and the styling gate below reads it.
//
// The BUCKET question — can an inline run express this payload? — lives in
// PseudoTextBridge, exactly as it does on iOS.

import com.styleconverter.runtime.PropertyTracker
import com.styleconverter.runtime.core.ir.IRComponent
import com.styleconverter.runtime.core.ir.IRLog
import kotlinx.serialization.json.JsonObject

/** Log tag for this fold's named refusals. */
private const val TAG = "PseudoTextFold"

object PseudoTextFold {

    /**
     * The fold's outcome: the (possibly rewritten) component plus whether
     * the `pseudos.after` bucket was CONSUMED here. `afterFolded = true`
     * obliges the caller to keep ContentApplier's wrapper off that bucket.
     */
    data class Folded(val component: IRComponent, val afterFolded: Boolean)

    /**
     * Fold the component's baked `::after` text onto the END of its own
     * text, per CSS 2.1 §12.1 order (element content, then ::after).
     *
     * Identity — the SAME instance back, `afterFolded = false` — for every
     * component this fold does not claim, which is every component of the
     * committed 327-fixture baseline corpus (none carries `pseudos`) and
     * every before-only bucket. Committed captures are byte-stable by
     * construction.
     *
     * ORDERING GUARDS, and their corpus weight. Two refusals below exist
     * only to keep css-pseudo-4 §4.1's paint order (::before, the element's
     * content, ::after): the `children` guard, and the box-shaped-`::before`
     * guard that follows it. Census over all 1435 `wave49-final`
     * per-test-ir documents: ONE component carries a box-shaped `::before`
     * (css-pseudo/before-as-flex-container) and it has no `::after`, so
     * ZERO documents reach the second guard — it moves no cell and exists
     * so the inversion cannot appear the moment a bucket does.
     */
    fun resolve(component: IRComponent): Folded {
        // The overwhelmingly common path: no generated content at all.
        val pseudos = component.pseudos ?: return Folded(component, false)
        // Root-scope buckets belong to RootPseudoBox (wave-28 lane PG):
        // that family is `content:"" + width/height/background`, a BOX and
        // not a text run, and folding here would double-render it. Same
        // guard direction as RootPseudoBox's banner, from the other shore.
        if (component.role == "body-root") return Folded(component, false)
        // No `::after` payload → nothing for this file to do. (`before` and
        // `marker` are deliberately untouched: the wrapper and the list
        // marker path own them — see the SCOPE note in the file banner.)
        val bucket = pseudos["after"] as? JsonObject ?: return Folded(component, false)
        // A bucket that generates a BLOCK-LEVEL BOX is PseudoBoxFold's, via
        // the SAME claim function PseudoBucketExtractor consults — so the
        // one payload can never render through two paths.
        if (PseudoGeneratedBox.claim(bucket, "after", component.role) != null) {
            return Folded(component, false)
        }
        // `meta.runs` is AUTHORITATIVE over `_text` (schema/spec/03-children.md
        // §4.1): when a run plan resolves, the renderer paints the listed
        // entries and drops the text slot, so a fold into `_text` would
        // silently vanish. Refuse and name it — never a silent fallthrough.
        if (component.runs != null) {
            PropertyTracker.markUnhandled("PseudoText::runs")
            IRLog.warn(TAG, "component carries meta.runs — ::after text not folded (runs own the content slot)")
            return Folded(component, false)
        }
        // css-pseudo-4 §4.1 places ::after AFTER the element's children. A
        // fold appends to the element's own TEXT, which paints BEFORE the
        // child stack — the wrong order — so a component with children keeps
        // the wrapper path it has today. Named, not silently mis-ordered.
        if (!component.children.isNullOrEmpty()) {
            PropertyTracker.markUnhandled("PseudoText::after-children")
            IRLog.warn(TAG, "::after on a component with children is not folded (it must trail the children)")
            return Folded(component, false)
        }
        // …and the SAME ordering rule from the other direction, which the
        // guard above cannot see (wave-50 skeptic S3). ComponentRenderer runs
        // THIS fold FIRST and PseudoBoxFold SECOND — `PseudoTextFold.resolve(
        // ContentsUnboxing.resolve(component))`, then `PseudoBoxFold.resolve(
        // pseudoTextFold.component)` — and PseudoBoxFold splices a claimed
        // `::before` box in as the host's FIRST CHILD. That child does not
        // exist yet when the children guard runs, so a host carrying a
        // box-shaped `::before` AND a text-shaped `::after` would fold the
        // `::after` glyphs into `_text`, which the renderer paints BEFORE the
        // child stack: the `::after` ink would land AHEAD of the `::before`
        // box — the exact inversion of css-pseudo-4 §4.1 (::before first, the
        // element's content, ::after last). One claim function, now three
        // consumers (PseudoBoxFold, PseudoBucketExtractor, here): refuse and
        // keep the wrapper path, which at least keeps the two runs in order.
        //
        // CORPUS CENSUS (all 1435 `tools/titan/runs/wave49-final/sections/
        // */per-test-ir` documents, this file's own predicate replayed with
        // the real PseudoGeneratedBox.claim): exactly ONE component in the
        // corpus carries a box-shaped `::before` at all —
        // css-pseudo/before-as-flex-container — and it carries NO `::after`,
        // so ZERO documents carry the both-buckets shape this guard refuses.
        // It cannot move a single corpus cell; it is a correctness guard
        // against the ordering inversion, not a score change.
        if (PseudoGeneratedBox.claim(
                pseudos["before"] as? JsonObject, "before", component.role) != null
        ) {
            PropertyTracker.markUnhandled("PseudoText::after-before-box")
            IRLog.warn(
                TAG,
                "::after is not folded under a box-shaped ::before " +
                    "(the fold would paint it ahead of the generated box)")
            return Folded(component, false)
        }
        // The declaration gate + the baked string. Null = nothing an inline
        // run can honestly express; every refusal inside is already named.
        val run = PseudoTextBridge.inlineRun(bucket, component)
            ?: return Folded(component, false)
        // css-cascade-5 §7.3 / css-pseudo-4 §2: `component._text` is ONE
        // uniform run, so the bucket's OWN styling may only apply when the
        // pseudo is the component's SOLE ink. Otherwise the host's text
        // would be repainted in the pseudo's colour/size — a wrong render
        // where today's wrapper at least keeps the two runs separate.
        val hostText = component._text ?: ""
        if (run.styling.isNotEmpty() && hostText.isNotEmpty()) {
            PropertyTracker.markUnhandled("PseudoText::styled-nonuniform-after")
            IRLog.warn(TAG, "::after styling cannot apply — the fold is not the component's sole text run")
            return Folded(component, false)
        }
        // The fold itself. `_text` bakes its own separators (" 1" / "1 "),
        // so plain concatenation is the whole job; the run is non-empty by
        // construction (inlineRun never returns an empty string).
        return Folded(
            component.copy(
                _text = hostText + run.text,
                // Appended LAST so the pseudo's own declarations win the
                // extractors' last-wins cascade, exactly as the author's
                // `::after { color: … }` beats the inherited value.
                properties = component.properties + run.styling
            ),
            afterFolded = true
        )
    }
}
